content/Installation/Rust.md

```diff
@@ -0,0 +1,63 @@
+---
+title: "Rust"
+date: 2019-11-29T15:26:15Z
+draft: false
+weight: 30
+---
+
+Getting started in Rust is easy, if you are using Linux. Otherwise, you probably need to wait for our next implementation, which will support Windows and Mac too.  
+We have published `enzyme` on [crates.io](https://crates.io/crates/enzyme).
+
+So after getting the following dependencies:
+
+- git
+- ninja
+- cmake
+- libssl-dev
+- libclang-dev
+- Rust (rustup) and any nightly toolchain 
+- ~10GB free storage in $HOME/.cache
+
+you can run
+```sh
+cargo install enzyme
+```
+
+This should only take a minute and give you an enzyme-install binary which you can use to build rustc/llvm/enzyme and a bunch of extra things, which might come helpful.  Now you can start the build process by calling
+```bash
+enzyme-install
+```
+It will build everything inside of `$HOME/.cache`, so please have around 10GB available there.
+Also take care, the build process takes around one hour on a recent 8 core cpu.
```

Comment:
If I already have a `LLVMEnzyme-13.so` built, is there a way to use it from Rust without this time-consuming step? Besides compilation cost, I'm interested in using Enzyme from C and Rust within the same application.

Reply:
@ZuseZ4, if I recall the issue here is not compiling Enzyme, but compiling Rust's LLVM, yes?

Also note that these instructions are aimed at the current work in progress nicer language integration (e.g. fewer compiler flag hacks + syntactic sugar) of Enzyme within Rust (see https://github.com/rust-ml/oxide-enzyme).

Using an existing Enzyme build (for a rust-compatible LLVM version) presumably would still work by either manually using the LLVM output of rust, or say using LLD as here (https://internals.rust-lang.org/t/automatic-differentiation-differential-programming-via-llvm/13188/7)

Reply:
Fair question. I was already wondering if I should add a note about just that while writing. The short version is maybe. 

As mentioned by @wsmoses, the main reason of having the current build procedure is stability. Each Rustc is only guaranteed to be compile-able against one specific LLVM commit (not even an LLVM release). So we could invest time to speed up the getting started process, but at the cost of introducing more bugs due to incompatibilities. Also, our main goal is a proper rustc integration, which is required due to some other issues. Once that's achieved we will archive the entire enzyme_build archive, so all optimizations here are only of limited duration. If you care about using C and rust together, 
I would just recommend you to use the clang version which I build along with rustc. This way Enzyme, Clang, rustc are
all based on the same LLVM version, which is probably the most reliable setup that you can get. That's what I'm using right now for some C++ code too. You will find clang in `/home/$USER/.cache/enzyme/rustc-1.57.0-src/build/x86_64-unknown-linux-gnu/llvm/bin/clang`. It might cost you two cpu-hrs, but it will probably safe you more hours of debugging/developing.

However, if you prefer to test your existing Enzyme version together with an official rustc, here is some background /history:
I added a flag to rustc, which allows loading LLVM plugins. This flag will only be available if you build rustc from source, as it is done by my script. Having this plugin interface has two purposes:
1) we can run some specific Enzyme optimizations before AD
2) We could use the __enzyme_ notation, as it is done in the C++ examples.

However: we moved away from the __enzyme notation and we haven't added the usage of the enzyme optimizations yet,
so you should theoretically be able to use some other rustc. You can do that by replacing `+enzyme` with `+nightly`
when building a crate. I just tested it an it fails due to having `__rust_probestack` defined more than once. I handle this [here](https://github.com/rust-ml/oxide-enzyme/blob/68a1f1eec636eeb1a07f273660abd72a9d5f7eba/src/lib.rs#L548), 
so you could try to debug why it's working with the enzyme, but not with the nightly toolchain.
Finally you could try to build rustc based on your existing LLVM Version, but as mentioned, rustc devs don't offer any
guarantees about that working. Also, building rustc-stage2 even without llvm takes a descent amount of time.

So yes, you might be able to sip this time-consuming step, I just really don't know if it's recommendable atm.
Either way, just let me know what you prefer, and I can try to help with that.

Reply:
Thanks for all these details. I'll just follow the procedure for now, though I hope it'll be possible to upstream what is necessary.

I'd be okay using one build of Enzyme with release clang for differentiating C and another built internally for rustc. Enzyme builds pretty fast and distribution will be a hassle if one has to always match. (For example, users may build our libraries on an HPC machine with vendor MPI set up to use their version of clang. Building rustc+llvm from source and then using environment variables to make the vendor MPI use our special clang will be an error-prone user experience.)

One follow-up: what's the process for incremental rebuilds instead of the released crate, say upon pulling Enzyme or if one is hacking on Enzyme?

Reply:
So I just pushed a new version to crates.io yesterday, which moved to the latest enzyme release 0.0.25. I've added some minimal logic to recognize whether enzyme or rustc are outdated, and only build what is needed.

If you made some changes to enzyme locally and want to test them, I recommend to just build it against the llvm from the rustc-1.57 build.
If you place your enzyme repo folder inside of '$HOME/.cache/enzyme' you can then update the ENZYME_VER in oxide-enzyme/src/lib.rs to whatever name you used for your folder.
You also need to adjust the LD_LIBRARY_PATH, but then you should be ready. I will try to add this also to the README before publishing the next enzyme version on crates.io.